File: scripts/extract_decryption_key_nky.py
#!/usr/bin/env python3

import base64
import subprocess
import sys
import tempfile


def print_usage():
    print("""
extract_decryption_key_nky.py <path_to_nky_file> <path_to_og_private_key>
    
    path_to_nky_file:           Path to the decryption key nky.
    path_to_og_private_key:     Path to the original private key that is NOT embedded.
    
""")


def run(file_path: str, og_private_key_path: str):
    try:
        with open(file_path, "rb") as f:
            decoded = base64.decodebytes(f.read())

            key_ciphertext = decoded[:256]
            safeword = decoded[256:259]
            iv = decoded[259:275]
            private_key_ciphertext = decoded[275:]

            if safeword.decode('utf-8') != "H4k":
                sys.exit("Safeword does not match. Probably corrupted key.")

            print(f"key_ciphertext => {list(key_ciphertext)}")
            print(f"safeword => {list(safeword)}")
            print(f"iv => {list(iv)}")
            print(f"private_key_ciphertext => {list(private_key_ciphertext)}")
            print(f"key_ciphertext => {key_ciphertext}")

            with tempfile.NamedTemporaryFile(delete=False) as tmp:
                print(tmp.name)
                tmp.write(key_ciphertext)

                decrypt_output = subprocess.run(
                    ["openssl", "rsautl", "-decrypt", "-oaep", "-in", tmp.name, "-inkey", og_private_key_path],
                    capture_output=True,
                    text=True,
                    check=True,
                )

                print(f"decrypt_output: {decrypt_output.stdout}")

    except FileNotFoundError:
        sys.exit("File was not found")


if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Please provide the correct arguments")
        print_usage()
        sys.exit()

    run(sys.argv[1], sys.argv[2])

File: scripts/extract_decryption_key_nky.js
#!/usr/bin/env nodejs

const crypto = require('crypto');
const path = require('path');
const fsPromises = require('fs').promises;

function printUsage() {
  console.log(`
  extract_decryption_key_nky.js <path_to_nky_file> <path_to_og_private_key>
  
    path_to_nky_file:           Path to the decryption key nky.
    path_to_og_keys_dir:        Path to the original keys: og_private.pem and pg_public.pem.
  
  `)
}

async function testKeypair(ogPrivateKey, ogPublicKey) {
  const plaintext = Buffer.from('Hello world!', 'utf8');

  const testCiphertext = crypto.publicEncrypt({
    key: ogPublicKey,
    padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
  }, plaintext);

  const decrypted = crypto.privateDecrypt({
    key: ogPrivateKey,
    padding: crypto.constants.RSA_PKCS1_OAEP_PADDING,
  }, testCiphertext);

  if (Buffer.compare(plaintext, decrypted) !== 0) {
    return Promise.reject(new Error('Key pair test failed'));
  }
}

async function parseNkyFile(nkyFilePath) {
  const encoded = await fsPromises.readFile(nkyFilePath);
  const decoded = Buffer.from(encoded.toString(), 'base64');

  const keyCiphertext = decoded.slice(0, 256);
  const safeword = decoded.slice(256, 259);
  const iv = decoded.slice(259, 275);
  const privateKeyCiphertext = decoded.slice(275);

  if (safeword.toString() !== 'H4k') {
    return Promise.reject(new Error('Safeword does not match'));
  }

  /*console.log('keyCiphertext =>', [...keyCiphertext]);
  console.log('safeword =>', [...safeword]);
  console.log('iv =>', [...iv]);
  console.log('privateKeyCiphertext =>', [...privateKeyCiphertext]);*/

  return Promise.resolve({
    keyCiphertext,
    safeword,
    iv,
    privateKeyCiphertext,
  });
}

async function run() {
  if (process.argv.length !== 4) {
    console.error('Please provide the correct arguments');
    printUsage();
    process.exit(1);
  }

  let ogPrivateKey = await fsPromises.readFile(path.join(process.argv[3], 'og_private.pem'));
  let ogPublicKey = await fsPromises.readFile(path.join(process.argv[3], 'og_public.pem'));

  try {
    const {
      keyCiphertext,
      safeword,
      iv,
      privateKeyCiphertext,
    } = await parseNkyFile(process.argv[2]);

    if (safeword.toString() !== 'H4k') {
      console.error('Safword check failed');
      process.exit(1);
    }

    await testKeypair(ogPrivateKey, ogPublicKey);

    const key = crypto.privateDecrypt(ogPrivateKey, keyCiphertext);
    const decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
    const privateKey = decipher.update(privateKeyCiphertext, null, 'utf8');

    console.log((new Buffer(privateKey)).toString('base64'));
  } catch (e) {
    console.error(e);
    process.exit(1);
  }


}

run();
